apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,485 @@
+      return;
+    }
+
+    this._bluetooth = window.navigator.mozBluetooth;
+
+    // listen states when the adapter is available
+    window.addEventListener('bluetooth-available',
+      this._adapterAvailableHandler.bind(this));
+
+    this._initDefaultAdapter();
+
+    this._bluetooth.addEventListener('attributechanged',
+      this._btManagerHandler.bind(this));
+
+    // decouple enable/disable bluetooth function from other system part
+    window.addEventListener('request-enable-bluetooth',
+      this._requestEnableHandler.bind(this));
+    window.addEventListener('request-disable-bluetooth',
+      this._requestDisableHandler.bind(this));
+
+    // relay bluetooth transfer system messages
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-start',
+      this._oppTransferStartHandler.bind(this));
+    navigator.mozSetMessageHandler('bluetooth-opp-transfer-complete',
+      this._oppTransferCompleteHandler.bind(this));
+
+    // expose isEnabled function to Service.query
+    Service.registerState('isEnabled', this);
+
+    // handle statusbar icons
+    LazyLoader.load(['js/bluetooth_icon.js',
+                     'js/bluetooth_transfer_icon.js',
+                     'js/bluetooth_headphone_icon.js'], function() {
+      this.icon = new BluetoothIcon(this);
+      this.icon.start();
+      this.transferIcon = new BluetoothTransferIcon(this);
+      this.transferIcon.start();
+      this.headphoneIcon = new BluetoothHeadphoneIcon(this);
+      this.headphoneIcon.start();
+    }.bind(this)).catch(function(err) {
+      console.error(err);
+    });
+  },
+
+  /**
+   * Generally bluetooth adapter is already available when start
+   * this module and no related event will be received.
+   * so we try to get the adapter then listen to mozBluetooth change
+   * to prevent race condition
+   *
+   * @private
+   */
+  _initDefaultAdapter: function() {
+    this._debug('init bluetooth adapter');
+    this._adapter = this._bluetooth.defaultAdapter;
+    if (this._adapter) {
+      // dispatch event when default adapter is available
+      this._dispatchAdapterState(true);
+
+      // dispatch default bluetooth enable state
+      this._isEnabled = (this._adapter.state === 'enabled');
+      this._dispatchEnableState();
+    } else {
+      // dispatch event when default adapter is not available
+      this._dispatchAdapterState(false);
+    }
+  },
+
+  /**
+   * Once adapter is available, listen to bluetooth state and
+   * the connect stat of supported profiles.
+   *
+   * we only support two use cases for bluetooth connection:
+   *   1. connecting with a headset
+   *   2. transfering a file to/from another device
+   * So we need to listen to corresponding events to know we are (aren't)
+   * connected, then summarize to an event and dispatch to StatusBar
+   *
+   * @private
+   */
+  _adapterAvailableHandler: function bt__adapterAvailableHandler() {
+    if (this._adapter) {
```

Comment:
This check seems not required because the handler is called only when the adapter is available.